_episodes/l4_04_parameterized_tests.md

```diff
@@ -0,0 +1,353 @@
+class EmployeeTestParameterised : public::TestWithParam<TestValues> {
+    public:
+        Employee employee{"John", 25, 8000, 3, 2000};
+};
+
+// Create an array of values (of type TestValues) to be injected into the test.
+TestValues values[] = {
+    TestValues{5, 2000},
+    TestValues{15, 3000}
+};
+
+//Test net bonus works fine for different number of years.
+TEST_P(EmployeeTestParameterised, NetBonusIsCorrectForDifferentYears) {
+    TestValues current_test_case_value = GetParam();
+    employee.setNumberYearsEmployed(current_test_case_value.input);
+    EXPECT_EQ(employee.getNetBonus(), current_test_case_value.output);
+}
+
+// Instantiate the test case with the values array.
+INSTANTIATE_TEST_SUITE_P(NetBonusIsCorrectForDifferentYears, 
+                         EmployeeTestParameterised,
+                         ValuesIn(values));
+```
+
+On running the above file, we see the following output.
+
+```bash
+[==========] Running 2 tests from 1 test suite.
+[----------] Global test environment set-up.
+[----------] 2 tests from NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised
+[ RUN      ] NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised.NetBonusIsCorrectForDifferentYears/0
+[       OK ] NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised.NetBonusIsCorrectForDifferentYears/0 (0 ms)
+[ RUN      ] NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised.NetBonusIsCorrectForDifferentYears/1
+[       OK ] NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised.NetBonusIsCorrectForDifferentYears/1 (0 ms)
+[----------] 2 tests from NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised (0 ms total)
+
+[----------] Global test environment tear-down
+[==========] 2 tests from 1 test suite ran. (0 ms total)
+[  PASSED  ] 2 tests.
+```
+
+In this output, there are two things worth noting:-
+
+1. As expected, we are now running two tests as compared to just one in case of a for loop.
+2. The test name `NetBonusIsCorrectForDifferentYears/EmployeeTestParameterised.NetBonusIsCorrectForDifferentYears/0` is a combination of the following:-
+    - A Prefix `NetBonusIsCorrectForDifferentYears` coming from INSTANTIATE_TEST_SUITE_P.
+    - Parameterised class name `EmployeeTestParameterised` coming from the first argument of `TEST_P` macro.
+    - Test name `NetBonusIsCorrectForDifferentYears` coming from the second argument of `TEST_P` macro.
+    - Finally, the iteration number.
+
+With this parameterised test, we were able to solve the issues that we were discussing above. However, in doing so, we changed the test fixture and converted it to use `TEST_P` macro. Our previous tests based on `TEST_F` macro will not work anymore as it has been replaced. The important question is: What shall we do so that we can still keep all our useful tests from test fixtures while still being able to add parameterised test. The solution is to combine test fixture with parameterised test and the next subsection explains that.
```

Comment:
```suggestion
With this parameterised test, we were able to solve the issues that we were discussing above. However, in doing so, we changed the test fixture and converted it to use `TEST_P` macro. Our previous tests based on `TEST_F` macro will not work anymore as it has been replaced. The important question is: What shall we do so that we can still keep all our useful tests from test fixtures while still being able to add parameterised test? The solution is to combine test fixtures with parameterised tests and the next subsection explains that.
```